engine/runtime-integration-tests/src/test/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNodeMultiValueTest.java

```diff
@@ -0,0 +1,150 @@
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.library.dispatch.TypeOfNode;
+import org.enso.interpreter.test.ValuesGenerator;
+import org.enso.test.utils.ContextUtils;
+import org.enso.test.utils.TestRootNode;
+import org.graalvm.polyglot.Context;
+import org.graalvm.polyglot.Value;
+import org.junit.AfterClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TypeOfNodeMultiValueTest {
+  @Parameterized.Parameter(0)
+  public Object value;
+
+  @Parameterized.Parameter(1)
+  public String type;
+
+  @Parameterized.Parameter(2)
+  public int typeIndex;
+
+  private static Context ctx;
+
+  private static Context ctx() {
+    if (ctx == null) {
+      ctx = ContextUtils.defaultContextBuilder().build();
+    }
+    return ctx;
+  }
+
+  @Parameterized.Parameters
+  public static Object[][] allPossibleEnsoInterpreterValues() throws Exception {
+    var g = ValuesGenerator.create(ctx());
+    var typeOf =
+        ContextUtils.evalModule(
+            ctx(),
+            """
+    from Standard.Base import all
+
+    typ obj = Meta.type_of obj
+    main = typ
+    """);
+    var data = new ArrayList<Object[]>();
+    for (var polyValue : g.allValues()) {
+      registerValue(g, typeOf, polyValue, data);
+    }
+    return data.toArray(new Object[0][]);
+  }
+
+  private static void registerValue(
+      ValuesGenerator g, Value typeOf, Value polyValue, ArrayList<Object[]> data) {
+    var t = typeOf.execute(polyValue);
+    if (!polyValue.isNull()) {
+      assertTrue("Type of " + polyValue + " is " + t, t.isMetaObject());
+      var rawValue = ContextUtils.unwrapValue(ctx(), polyValue);
+      var rawType = ContextUtils.unwrapValue(ctx(), t);
+      if (rawType instanceof Type type) {
+        var singleMultiValue = EnsoMultiValue.create(new Type[] {type}, new Object[] {rawValue});
+        var n = t.getMetaSimpleName();
+        data.add(new Object[] {singleMultiValue, n, 0});
+        var rawInt = (Type) ContextUtils.unwrapValue(ctx(), g.typeInteger());
+        var secondMultiValue =
+            EnsoMultiValue.create(new Type[] {rawInt, type}, new Object[] {5L, rawValue});
+        data.add(new Object[] {secondMultiValue, n, 1});
+        var firstMultiValue =
+            EnsoMultiValue.create(new Type[] {type, rawInt}, new Object[] {rawValue, 6L});
+        data.add(new Object[] {firstMultiValue, n, 0});
+      } else {
+        if (!t.isHostObject()) {
+          data.add(new Object[] {rawValue, null, -1});
+        }
+      }
+    }
+  }
+
+  @AfterClass
+  public static void disposeCtx() throws Exception {
+    if (ctx != null) {
+      ctx.close();
+      ctx = null;
+    }
+  }
+
+  @Test
+  public void typeOfCheck() {
+    assertType(value, type, typeIndex, false);
+  }
+
+  @Test
+  public void typeOfCheckAfterPriming() {
+    assertType(value, type, typeIndex, true);
+  }
+
```

Comment:
Probably worth creating the `TestRootNode` and `TypeOfNode` in some `@BeforeClass` initializer. You generally want to pass as many arguments to `TypeOfNode.execute` as possible so that many specializations will get instantiated. You are recreating the node with blank state before every `execute` method call. You could also get rid of `priming`.

Reply:
`priming` was a way to trigger (originally broken) specialization in the original `TypeOfNodeTest`.

> so that many specializations will get instantiated

that's a bit uncontrolled process! The specializations are created randomly (without my brain thinking about them) and the result is two sword: it may hide some problems, or it may show some problems. Just like creating new node for every operation.

Reply:
While the original `TypeOfNodeTest` may be complicated with `priming` for a reason (there was a bug the test exactly triggers). There is no need to complicate the new test. Simplified in 55bea9b60557b9d1879cafa2846f8f2e9fe00c92